File: examples/ruby/spec/elements/finders_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe 'Element Finders' do
  let(:driver) { start_session }
  let(:locators_page) { 'https://www.selenium.dev/selenium/web/locators_tests/locators.html' }

  it 'finds the first matching element' do
    driver.navigate.to locators_page
    first_input = driver.find_element(class: 'information')

    expect(first_input.attribute('id')).to eq('fname')
  end

  it 'uses a subset of the dom to find an element' do
    driver.navigate.to locators_page
    form = driver.find_element(tag_name: 'form')
    input_element = form.find_element(class: 'information')

    expect(input_element.attribute('id')).to eq('fname')
  end

  it 'uses an optimized locator' do
    driver.navigate.to locators_page
    input_element = driver.find_element(css: 'form .information')

    expect(input_element.attribute('id')).to eq('fname')
  end

  it 'finds all matching elements' do
    driver.navigate.to locators_page
    inputs = driver.find_elements(tag_name: 'input')

    expect(inputs.size).to be > 1
  end

  # rubocop:disable RSpec/Output
  it 'gets an element from a collection' do
    driver.navigate.to locators_page
    elements = driver.find_elements(tag_name: 'p')
    elements.each { |e| puts "Paragraph text:#{e.text}" }

    expect(elements).not_to be_empty
  end

  it 'finds element from element' do
    driver.navigate.to locators_page
    form = driver.find_element(tag_name: 'form')
    elements = form.find_elements(tag_name: 'input')
    elements.each { |e| puts e.attribute('value') }

    expect(elements).not_to be_empty
  end
  # rubocop:enable RSpec/Output

  it 'finds the active element' do
    driver.navigate.to locators_page
    driver.find_element(css: '#fname').send_keys('webElement')
    attr = driver.switch_to.active_element.attribute('name')

    expect(attr).to eq('fname')
  end
end

File: examples/javascript/test/elements/finders.spec.js
const {Builder, By} = require('selenium-webdriver');
const assert = require('assert');

const LOCATORS_PAGE = 'https://www.selenium.dev/selenium/web/locators_tests/locators.html';

describe('Finders', function () {
  it('finds the first matching element', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const firstInput = await driver.findElement(By.className('information'));

    assert.equal(await firstInput.getAttribute('id'), 'fname');
    await driver.quit();
  });

  it('finds an element within a subset of the DOM', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const form = await driver.findElement(By.tagName('form'));
    const input = await form.findElement(By.className('information'));

    assert.equal(await input.getAttribute('id'), 'fname');
    await driver.quit();
  });

  it('uses an optimized locator', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const input = await driver.findElement(By.css('form .information'));

    assert.equal(await input.getAttribute('id'), 'fname');
    await driver.quit();
  });

  it('finds all matching elements', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const inputs = await driver.findElements(By.tagName('input'));

    assert.ok(inputs.length > 1);
    await driver.quit();
  });

  it('gets an element from a collection', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const elements = await driver.findElements(By.tagName('p'));
    for (const element of elements) {
      console.log('Paragraph text:' + await element.getText());
    }

    assert.ok(elements.length > 0);
    await driver.quit();
  });

  it('finds elements from an element', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    const form = await driver.findElement(By.css('form'));
    const elements = await form.findElements(By.css('input'));
    for (const e of elements) {
      console.log(await e.getAttribute('value'));
    }

    assert.ok(elements.length > 0);
    await driver.quit();
  });

  it('gets the active element', async function () {
    let driver = await new Builder().forBrowser('chrome').build();
    await driver.get(LOCATORS_PAGE);
    await driver.findElement(By.css('#fname')).sendKeys('webElement');
    const attr = await driver.switchTo().activeElement().getAttribute('name');

    assert.equal(attr, 'fname');
    await driver.quit();
  });
});
